services/src/main/java/io/grpc/services/HealthCheckingLoadBalancerFactory.java

```diff
@@ -0,0 +1,399 @@
+                  handleStreamClosed(status);
+                }
+              });
+          }
+        };
+    private final LoadBalancer delegate;
+    private final SynchronizationContext syncContext;
+    private final ScheduledExecutorService timerService;
+
+    private Subchannel subchannel;
+
+    // Set when the RPC is started. Cleared when the RPC is fully closed.
+    @Nullable
+    private ClientCall<HealthCheckRequest, HealthCheckResponse> activeCall;
+
+    // States about the most recently started RPC
+    private String lastCallServiceName;
+    private long lastCallStartNanos;
+    private boolean lastCallHasResponded;
+
+    // The service name should be used for health checking
+    private String serviceName;
+    private BackoffPolicy backoffPolicy;
+    // The state from the underlying Subchannel
+    private ConnectivityStateInfo rawState = ConnectivityStateInfo.forNonError(IDLE);
+    // The state concluded from health checking
+    private ConnectivityStateInfo concludedState = ConnectivityStateInfo.forNonError(IDLE);
+    // true if a health check stream should be kept.  Either there is an active RPC, or a retry is
+    // pending.
+    private boolean running;
+    // true if server returned UNIMPLEMENTED
+    private boolean disabled;
+    private ScheduledHandle retryTimer;
+
+    HealthCheckState(
+        LoadBalancer delegate, SynchronizationContext syncContext,
+        ScheduledExecutorService timerService) {
+      this.delegate = checkNotNull(delegate, "delegate");
+      this.syncContext = checkNotNull(syncContext, "syncContext");
+      this.timerService = checkNotNull(timerService, "timerService");
+    }
+
+    void handleResponse(HealthCheckResponse response) {
+      backoffPolicy = null;
+      lastCallHasResponded = true;
+      if (!running) {
+        return;
+      }
+      if (Objects.equal(response.getStatus(), ServingStatus.SERVING)) {
+        gotoState(ConnectivityStateInfo.forNonError(READY));
+      } else {
+        gotoState(
+            ConnectivityStateInfo.forTransientFailure(
+                Status.UNAVAILABLE.withDescription(
+                    "Health-check service responded "
+                    + response.getStatus() + " for '" + lastCallServiceName + "'")));
+      }
+      activeCall.request(1);
+    }
+
+    void handleStreamClosed(Status status) {
+      activeCall = null;
+      if (Objects.equal(status.getCode(), Code.UNIMPLEMENTED)) {
+        // TODO(zhangkun83): record this to channel tracer
+        disabled = true;
+        gotoState(rawState);
+        return;
+      }
+      if (running) {
+        long delayNanos = 0;
+        if (Objects.equal(lastCallServiceName, serviceName)) {
+          // Service name has not just changed.
+          gotoState(
+              ConnectivityStateInfo.forTransientFailure(
+                  Status.UNAVAILABLE.withDescription(
+                      "Health-check stream was erroneously closed with "
```

Comment:
erroneous means "wrong" or "incorrect." Should it be "Health-check stream failed with " ?

Reply:
It also handles the case that the stream is closed without error. I have reworded it to "unexpectedly closed"